### PR TITLE
Refactor test

### LINKED ISSUE #602: Avoid spwan command with rm
It spawns too many commands on removing files.

Why do we just use temp dir for this? 
else the CI server we have a lot of file operations.

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/manager.rs#L22-L33

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/bitcoin_manager.rs#L55
https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/cockroach_manager.rs#L49
https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/kld_manager.rs#L96
https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/electrs_manager.rs#L45

### LINKED ISSUE #604: Fix improper macros in test
We do not get any benefits from macros here, but it will slow down the test cases, each test case needs to expand the macro in compiling time.

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/bitcoin_manager.rs#L95-L112

There are more shortages using macro here.

- The `CARGO_BIN_EXE_kld`(in `kld!` macro) is in a different crate under compiling time.
https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/kld_manager.rs#L174C19-L174C36

- The `mut` variable is hidden and not passed with `mut` (the `setting` is changed in `bitcoin!`)
https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/kld/tests/bitcoind.rs#L12-L13

- The async function is hidden inside the macro

The following are macros with the same issue.

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/cockroach_manager.rs#L104-L119

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/electrs_manager.rs#L72-L88

https://github.com/kuutamolabs/lightning-knd/blob/0c70121aba785adb4bc2d7d08171c5f42dcabffb/test-utils/src/kld_manager.rs#L169-L188

### PR DESCRIPTION
**Pull request by yanganto**
fix #602
fix #604

### CODE REVIEW

File: kld/tests/smoke/start.rs [hunk 1]
```diff
@@ -17,30 +17,51 @@ use kld::{
...
+    let electrs_0 = ElectrsManager::new(&tmp_dir, &bitcoin, &mut settings_0).await?;
+    let kld_0 = KldManager::new(
+        &tmp_dir,
+        env!("CARGO_BIN_EXE_kld"),
+        &bitcoin,
+        &cockroach,
+        &electrs_0,
+        &mut settings_0,
+    )
+    .await?;
```
Comment (JosephGoulden) [inline kld/tests/smoke/start.rs:48]: The macros help by removing this boilerplate code but I can see how they are confusing. In the future we could have many more of these types of end to end test. We can find another way of hiding the setup code.

File: kld/tests/api/rest.rs [hunk 2]
```diff
@@ -903,6 +904,7 @@ pub struct TestContext {
     pub settings: Settings,
     admin_macaroon: Vec<u8>,
     readonly_macaroon: Vec<u8>,
+    _tmp_dir: TempDir,
```
Comment (JosephGoulden) [inline kld/tests/api/rest.rs:907]: why the underscore?

**Review by JosephGoulden — CHANGES_REQUESTED** [review]
Good idea using tempdir!

File: test-utils/src/cockroach_manager.rs [hunk 3]
```diff
@@ -9,17 +9,31 @@ use async_trait::async_trait;
...
-impl CockroachManager {
+impl<'a> CockroachManager<'a> {
+    pub async fn new(
+        output_dir: &'a TempDir,
+        settings: &mut Settings,
+    ) -> Result<CockroachManager<'a>> {
+        let mut cockroach = CockroachManager::test_cockroach(output_dir, &settings.node_id)?;
+
+        // XXX
+        // Why we need an extra memory copy here?
```
[the remark this answers is not in the document]
Comment (yanganto) [reply test-utils/src/cockroach_manager.rs:31]: It is a different thing. 
This is wrong typing in the `settings.database_port`, it should not be a string here.

**Review by JosephGoulden — COMMENTED** [review]
This still occasionally fails with 'bad file descriptor' on CI and on my desktop. 

Also I remembered that we use the target dir for test output so that it is easy to read the logs when the test fails. We can't do that with tempdir as it deletes them, so I don't think we can merge this.

**Comment by yanganto** [comment]
> This still occasionally fails with 'bad file descriptor' on CI and on my desktop.

Even avoiding spawning multiple rm commands does not solve this issue, we should do it correctly.

> Also I remembered that we use the target dir for test output so that it is easy to read the logs when the test fails. We can't do that with tempdir as it deletes them, so I don't think we can merge this.

It is best practice to correct setup and tear down test cases.  If you want to keep the log from a test, you should use [keep](https://docs.rs/tempfile/3.8.0/tempfile/struct.NamedTempFile.html#method.keep), [persist](https://docs.rs/tempfile/3.8.0/tempfile/struct.NamedTempFile.html#method.persist) or [persist_noclobber](https://docs.rs/tempfile/3.8.0/tempfile/struct.NamedTempFile.html#method.persist_noclobber) to debug and develop the test case.  When running a regular test on a CI server or other developers' machine, these tests should not leave any file on the machine without any prompts.

**Comment by yanganto** [comment]
FYI,

https://www.browserstack.com/guide/how-to-write-test-cases#toc5

> Create self-cleaning test cases. Test cases must revert the [Test Environment](https://www.browserstack.com/guide/what-is-test-environment) to a pristine, pre-test state. Tests should not leave remnants of themselves in the environment when completed. This is an integral element of Configuration Management. To understand more in-depth: [What is Configuration Management in DevOps?](https://www.browserstack.com/guide/configuration-management-in-devops)

**Comment by yanganto** [comment]
Alright, you can set any path to store the artifacts from test cases like this command.

`KEEP_TEST_ARTIFACTS_IN=/tmp/kld-test cargo test test_start`

Just kindly note that the working folder is hidden (created by the library) under `KEEP_TEST_ARTIFACTS_IN`.

**Comment by yanganto** [comment]
After today's meeting, the PR splits, and the keeping test artifacts will in #619

**Pull request merged**